headerCheck/rawHeaderHandler.go

```diff
@@ -0,0 +1,189 @@
+package headerCheck
+
+import (
+	"context"
+	"encoding/hex"
+
+	"github.com/ElrondNetwork/elrond-go-core/core"
+	"github.com/ElrondNetwork/elrond-go-core/core/check"
+	"github.com/ElrondNetwork/elrond-go-core/data/block"
+	"github.com/ElrondNetwork/elrond-go-core/marshal"
+	"github.com/ElrondNetwork/elrond-go/state"
+)
+
+type rawHeaderHandler struct {
+	proxy      Proxy
+	marshaller marshal.Marshalizer
+}
+
+// NewRawHeaderHandler will create a new rawHeaderHandler instance
+func NewRawHeaderHandler(proxy Proxy, marshaller marshal.Marshalizer) (*rawHeaderHandler, error) {
+	if check.IfNil(proxy) {
+		return nil, ErrNilProxy
+	}
+	if check.IfNil(marshaller) {
+		return nil, ErrNilMarshaller
+	}
+
+	return &rawHeaderHandler{
+		proxy:      proxy,
+		marshaller: marshaller,
+	}, nil
+}
+
+// GetMetaBlockByHash will return the MetaBlock based on the raw marshalized
+// data from proxy
+func (rh *rawHeaderHandler) GetMetaBlockByHash(ctx context.Context, hash string) (*block.MetaBlock, error) {
+	metaBlockBytes, err := rh.proxy.GetRawBlockByHash(ctx, core.MetachainShardId, hash)
+	if err != nil {
+		return nil, err
+	}
+
+	blockHeader := &block.MetaBlock{}
+	err = rh.marshaller.Unmarshal(blockHeader, metaBlockBytes)
+	if err != nil {
+		return nil, err
+	}
+
+	return blockHeader, nil
+}
+
+// GetShardBlockByHash will return the Header based on the raw marshalized data
+// from proxy
+func (rh *rawHeaderHandler) GetShardBlockByHash(ctx context.Context, shardId uint32, hash string) (*block.Header, error) {
+	metaBlockBytes, err := rh.proxy.GetRawBlockByHash(ctx, shardId, hash)
+	if err != nil {
+		return nil, err
+	}
+
+	blockHeader := &block.Header{}
+	err = rh.marshaller.Unmarshal(blockHeader, metaBlockBytes)
+	if err != nil {
+		return nil, err
+	}
+
+	return blockHeader, nil
+}
+
+// GetValidatorsInfoPerEpoch will return validators info based on start of
+// epoch metablock for a specific epoch
+func (rh *rawHeaderHandler) GetValidatorsInfoPerEpoch(ctx context.Context, epoch uint32) ([]*state.ShardValidatorInfo, []byte, error) {
+	lastStartOfEpochMetaBlock, err := rh.getLastStartOfEpochMetaBlock(ctx)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	metaBlock, randomness, err := rh.getMetaBlockAndRandomnessForEpoch(ctx, epoch, lastStartOfEpochMetaBlock)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	validatorsInfoPerEpoch, err := rh.getValidatorsInfo(ctx, metaBlock)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	return validatorsInfoPerEpoch, randomness, nil
+}
+
+func (rh *rawHeaderHandler) getMetaBlockAndRandomnessForEpoch(
+	ctx context.Context,
+	epoch uint32,
+	metaBlock *block.MetaBlock,
+) (*block.MetaBlock, []byte, error) {
+	var err error
+	randomness := metaBlock.GetPrevRandSeed()
+	currEpoch := metaBlock.GetEpoch()
+
+	for epoch <= currEpoch {
```

Comment:
please change this behavior as it is sub-optimal
We should have a dedicated node endpoint route that returns the marshaled current start of epoch metablock. All nodes have that block and all nodes can recover the last epoch's metablock from a single get from their storers.

Reply:
added dedicated endpoint to fetch start of epoch metablock from storage (on node)